phoenix-core/src/main/java/org/apache/phoenix/compile/QueryCompiler.java

```diff
@@ -147,21 +169,146 @@ public QueryCompiler(PhoenixStatement statement, SelectStatement select, ColumnR
+            ProjectedColumnExpression expression = new ProjectedColumnExpression(tbl.getColumns().get(i), tbl.getColumns(), i, tbl.getColumns().get(i).getExpressionStr());
+            projectedColumns.add(new ExpressionProjector(tbl.getColumns().get(i).getName().getString(), tbl.getName().getString(), expression, true));
+        }
+        RowProjector rowProjector = new RowProjector(projectedColumns, 100, true);
+        TupleProjector tupleProjector = new TupleProjector(rowProjector);
+        plan = new TupleProjectionPlan(plan, tupleProjector, null);
+        return plan;
+    }
+    private boolean containOrderBy(SelectStatement select) {
+        if (select.getOrderBy() != null && !select.getOrderBy().isEmpty()) 
+            return true;
+        else
+            return false;
+    }
+
+    private QueryPlan compileUnionAll(SelectStatement select) throws SQLException { 
+        List<SelectStatement> unionAllSelects = select.getSelects();
+        List<QueryPlan> plans = new ArrayList<QueryPlan>();
         StatementContext context = new StatementContext(statement, resolver, scan, sequenceManager);
-        if (select.isJoin()) {
-            select = JoinCompiler.optimize(statement, select, resolver);
-            if (this.select != select) {
-                ColumnResolver resolver = FromCompiler.getResolverForQuery(select, statement.getConnection());
-                context = new StatementContext(statement, resolver, scan, sequenceManager);
+
+        checkForOrderByLimitInUnionAllSelect(select);
+        QueryPlan plan = compileSelect(select);
+        plan = buildTupleProjectPlan(plan);
+        plans.add(plan);
+        OrderBy orderBy = OrderBy.EMPTY_ORDER_BY;
+        boolean containOrderBy = false;
+        int numSelects = unionAllSelects.size();
+        for (int i=0; i < numSelects; i++ ) {
+            if (i < numSelects-1)
+                checkForOrderByLimitInUnionAllSelect(unionAllSelects.get(i));
+            else if (i == numSelects-1) {
+                containOrderBy = containOrderBy(unionAllSelects.get(i));
             }
-            JoinTable joinTable = JoinCompiler.compile(statement, select, context.getResolver());
-            return compileJoinQuery(context, binds, joinTable, false, false, null);
-        } else {
-            return compileSingleQuery(context, select, binds, false, true);
+            plan = compileSelect(unionAllSelects.get(i));
+            if (containOrderBy) {
+                orderBy = plan.getOrderBy();
+                unionAllSelects.get(i).removeOrderBy();
+                plan = compileSelect(unionAllSelects.get(i));
+            } 
+            plan = buildTupleProjectPlan(plan);
+            plans.add(plan);
         }
+        checkProjectionNumAndTypes(plans);
+        plan =  new UnionPlan(context, plan.getStatement(), plan.getTableRef(), plan.getProjector(), plan.getLimit(), orderBy, parallelIteratorFactory, plan.getGroupBy(), plans);
```

Comment:
OrderBy is used by union all plan to get Order by information. Cannot be null if it has order by clause.

When building the RowProjector, which "select" you referred to here. There are many selects in union all. Not matter which one you choose, effectively you pick one of the selects to build the RowProjector. It is the problem I have mentioned. And that is why I pass down RowProjector to result sets.
RowProjector rowProjector = ProjectorCompiler.compile(context, select, groupBy, Collections.emptyList());